core/src/banking_stage/unprocessed_transaction_storage.rs

```diff
@@ -181,13 +174,18 @@ fn consume_scan_should_process_packet(
             payload
                 .message_hash_to_transaction
                 .remove(packet.message_hash());
-            ProcessingDecision::Never
-        } else if payload.account_locks.try_locking(message) {
-            payload.sanitized_transactions.push(sanitized_transaction);
-            ProcessingDecision::Now
-        } else {
-            ProcessingDecision::Later
+            return ProcessingDecision::Never;
         }
+
+        // Always take locks during batch creation.
+        // This prevents lower-priority transactions from taking locks
+        // needed by higher-priority txs that were skipped by this check.
+        if !payload.account_locks.take_locks(message) {
```

Comment:
if this returns false, this means that either the batch already has a transaction with these locks
OR
we didn't add a tx due to this check.

### Example

MI, assume 2 iterators (write-locks only for simplicity):
```
[(AB), (BC), (CD)]
```

first iteration:
```
- (AB) - selected.      WriteLockSet {A, B}
- (BC) - not selected.  WriteLockSet {A, B, C}
- (CD) - not selected.  WriteLockSet {A, B, C, D}
```

Even though `BC` cannot be selected, we add `C` to the set to prevent lower-priority txs locking `C` from doing so.

Reply:
If `(CD)` has higher priority, it should be selected, but won't due to lower-priority `(BC)` took lock on `C`; that's not what we want, is it?

Maybe needs two sets of lock status: one for locks taken; another for "want to take" with the highest priority?

Reply:
I should have been explicit; the buffer `[(AB), (BC), (CD)]` is in priority-order, so (BC) has the higher priority.

If `CD` had higher-priority then we'd have:
```
[(AB), (CD), (BC)]
```
first iteration:
```
- (AB) - selected.      WriteLockSet {A, B}
- (CD) - selected.      WriteLockSet {A, B, C, D}
- (BC) - not selected.  WriteLockSet {A, B, C, D}
```

Reply:
Oh right, I totally overlooked the prio ordering part. Yea, this makes sense then.